Module 2 Progress

britboy4321 opened the pull request.

hugh-emerson commented on app.py:
```diff
@@ -1,21 +1,165 @@
+    for todo in the_list:
+        superlist.append(todo["name"])                              # for each of the full cards in the list 
+
+# I tried all the below but literally nothing would work in order to get those card id's into different parts of the list, so next tried this:
```
Sorry, not sure what you wanted to achieve. I'm happy to discuss it if you had something particular in mind that you wanted to do but couldn't work out how to do it in Python.

britboy4321 commented:
Thankyou for showing such incredible patience in me, I really do
appreciate it.

I'm on holiday right now, back September when I will implement all of your
suggestions.  I'm trying my best!!!  I'm gonna nail it.

Your assumptions about my comments referring to older versions of my code
are all 100% true.  I'll fix it all :)

Thanks again,

Dave

On Mon, 24 Aug 2020, 6:26 pm Hugh Emerson, <notifications@github.com> wrote:

> *@hugh-emerson* requested changes on this pull request.
>
> You're almost there but can you please make the following changes:
>
>    - Tidy up app.py by deleting everything unnecessary
>    - Create an Item class - see my comment at the end of index()
>    - Technically, yes, your app can move todos to "Done", but only if the
>    user enters an id that isn't visible to them through the app. Can you try
>    following my suggestions for getting the button on each row working.
>
> If you ever feel like you're not making progress and don't know where to
> look next, then please do book a 30 minute 1 to 1 with one of us. It sounds
> like it would have saved you a lot of time.
> ------------------------------
>
> In .vscode/settings.json
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475701521>
> :
>
> > @@ -0,0 +1,3 @@
> +{
> +    "python.pythonPath": "c:\\Users\\Dave\\DevOps-Course-Starter\\env\\Scripts\\python.exe"
>
> Use a relative path instead of include C:\Users\Dave\DevOps-Course-Starter.
>
> Or equally, you could gitignore this file.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475708084>
> :
>
> > @@ -1,21 +1,165 @@
> +# SETUP INFO
>
> You have a *lot* of comments and commented out code in this file that
> should be deleted.
>
> Starting with line 1, does SETUP INFO really help you read this file?
>
> This file is 165 lines long, but should be more like 65.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475708190>
> :
>
> > -@app.route('/')
> +
> +#Set up variables we'll be using
> +
> +trellokey=os.environ["key"]            # get the secret key
> +trellotoken=os.environ["token"]         # get the secret token
> +cardsurl = "https://api.trello.com/1/cards"      # I could put the bit apart from 'cards' in some kind of 'parenturl' variable, but haven't!
> +acardsurl = "https://api.trello.com/1/cards/5f352898cdb592062bb71e23"                   # Used for hardcoded single card transfer from TO DO to DONE.
> +boardurl = "https://api.trello.com/1/boards/5f3528983d4fb244aae9f934/cards"             # The board ID is not a secret!
> +destinationlistid= "5f35289963b720014dd8a0ce"              # For the TO DO list and our destination for list transfer.  This is not a secret!  But is of course unique to me.  Found using postman API request,
> +listid = "5f352898dc8a8c31a0a1e439"                         # TO BE DONE list id - necessary for POST API CALL to add new items
> +donelistid = "5f3528981725711087e10339"                     # ACTUALLY DONE LIST ID
> +
> +# My code deals with a maximum of 100 cards.
> +
> +thislist=[]                                                                #Final list of ALL information in cards, nothing stripped out .. ready for ...
>
> Unused. Delete this line
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475708470>
> :
>
> > +
> +#Set up variables we'll be using
> +
> +trellokey=os.environ["key"]            # get the secret key
> +trellotoken=os.environ["token"]         # get the secret token
> +cardsurl = "https://api.trello.com/1/cards"      # I could put the bit apart from 'cards' in some kind of 'parenturl' variable, but haven't!
> +acardsurl = "https://api.trello.com/1/cards/5f352898cdb592062bb71e23"                   # Used for hardcoded single card transfer from TO DO to DONE.
> +boardurl = "https://api.trello.com/1/boards/5f3528983d4fb244aae9f934/cards"             # The board ID is not a secret!
> +destinationlistid= "5f35289963b720014dd8a0ce"              # For the TO DO list and our destination for list transfer.  This is not a secret!  But is of course unique to me.  Found using postman API request,
> +listid = "5f352898dc8a8c31a0a1e439"                         # TO BE DONE list id - necessary for POST API CALL to add new items
> +donelistid = "5f3528981725711087e10339"                     # ACTUALLY DONE LIST ID
> +
> +# My code deals with a maximum of 100 cards.
> +
> +thislist=[]                                                                #Final list of ALL information in cards, nothing stripped out .. ready for ...
> +superlist=[]                                                               #Final list of actual text in cards, with rest of information stripped out .. ready for HTML
>
> Why is this here? It's only used inside index(), so it should be declared
> inside index().
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475724042>
> :
>
> >  def index():
> -    Items=session.get_items()
> -   # session.add_item('A super entry')   # This works to add entries
> -    return render_template('index.html',passedItems=Items)
> +    # print ("Main application successfully refreshed")
> +    Items=session.get_items()
> +
> +# Trello GET for recieving all the cards here
> +# First, set up the query for the board
> +    query = {
> +        'key': trellokey,
> +        'token': trellotoken                        # That's all I need for this API call
> +    }
> +    response2 = requests.request(                    # Could have just used 'response' but needed a differentiator during coding.
>
> You didn't need the 2, the other "response" variables are in different
> scopes. But also, you've now ended up with two "response2" variables inside
> index().
>
> More importantly, the correct thing to do is not just stick a number on
> it, it's to use a meaningful name. What is this object more specifically?
> You could call it board_response to be consistent with your "boardurl"
> variable, or "cards_response", in that you made a request to get a list of
> cards and this is the response.
> ------------------------------
>
> In templates/index.html
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475728857>
> :
>
> >        </ul>
> +
> +
> +      <h2>Items from TRELLO To-DO list</h2>
> +      <ul class="list-group mb-4">
> +        <!--Literally no for loop would work.  After 4 hours of trying, It seems impossible to display a passed Python list as HTML list - so I had no choice but to do the below-->
>
> I'm confused by this comment. You do have a for loop displaying an html
> list.
>
> Did you forget to delete it after you managed to write the for loop?
>
> If you wanted something different, can you clarify what?
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475730138>
> :
>
> > +    # Desperate to try and turn the list into a dictionery so somehow I can get HTML to print it into different lines so I can add a button against them, but it just simply isn't going to work I can only get the whole list printed as a single line in the HTML which is of course useless.
> +
> +    # During module_1 a dict was returned that was in a format that somehow HTML seems to deal with.  No matter what I try I can't get 'superlist' into a format that simulates the format of the cookies list that works on the HTML.  It's too hard.  I've spent 4 hours on it.  It seems impossible with the info we have.
> +    # I'm having to very deep dive the 'enumerate' and indices operations but I can't spend lots more hours in this.
> +    #superlist2 = {key: [index for index, x in enumerate(superlist) if x == key] for key in superlist}.  All the below fails.
> +    #my_dict={
> +    #for key in superlist:
> +    #    indices = []
> +    #    for index,value in enumerate(superlist):
> +    #        if value == key:
> +    #            indices.append(index)
> +    #    my_dict[key] = indices
> +    #print(my_dict)
> +
> +    # print(superlist2)
>
> This whole comment block seems out of date? You are successfully
> displaying "superlist" in the HTML.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475732245>
> :
>
> > +# CHANGE LIST THAT A CARD IS IN.   THE MVP FOR THE SPEC IS I MOVE ANY CARD FROM TO DO INTO DOING.  So this is triggered by execution (or web page refresh) and hard-code moves one of my tasks to a different list.
> +# This PUT should be in a different method altogether but I've spent so many hours already that I have to move forward somehow.
> +#
> +    query = {
> +        'key': trellokey,
> +        'token': trellotoken,
> +        'idList': destinationlistid
> +    }
> +    response2 = requests.request(
> +         "PUT",
> +         acardsurl,
> +         params=query
> +     )
>
> This can be deleted. You already have a complete_item function that
> updates a card whenever you go to /complete_item.
> The lines declaring acardsurl and destinationlistid can also be deleted
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475734238>
> :
>
> > +    query = {
> +        'key': trellokey,
> +        'token': trellotoken                        # That's all I need for this API call
> +    }
> +    response2 = requests.request(                    # Could have just used 'response' but needed a differentiator during coding.
> +         "GET",
> +         boardurl,
> +         params=query
> +     )
> +
> + #   print(response2.text)                           # Just to check its actually being gotten!
> +
> +    #print(response2.text)
> +    #print(type(response2.text))                     # OK so response2.text is one massive long string.  Which is fairly useless
> +                                                    # It was a nightmare getting the bit of the JSON I wanted
> +    the_list = json.loads(response2.text)           # transfer the one long string into a LIST (Not a dictionary .. pretty useless now I'm gonna have to split it by hand). I REALLY with I knew how to get it into a dictionery, would have saved hours.
>
> Not sure what your comment mean by "I'm gonna have to split it by hand".
> the_list is a list of cards. Each card is a dictionary. Are these
> comments also out of date?
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475738596>
> :
>
> > +
> +    # During module_1 a dict was returned that was in a format that somehow HTML seems to deal with.  No matter what I try I can't get 'superlist' into a format that simulates the format of the cookies list that works on the HTML.  It's too hard.  I've spent 4 hours on it.  It seems impossible with the info we have.
> +    # I'm having to very deep dive the 'enumerate' and indices operations but I can't spend lots more hours in this.
> +    #superlist2 = {key: [index for index, x in enumerate(superlist) if x == key] for key in superlist}.  All the below fails.
> +    #my_dict={
> +    #for key in superlist:
> +    #    indices = []
> +    #    for index,value in enumerate(superlist):
> +    #        if value == key:
> +    #            indices.append(index)
> +    #    my_dict[key] = indices
> +    #print(my_dict)
> +
> +    # print(superlist2)
> +
> +    return render_template('index.html',passedItems=Items,todisplay=superlist)
>
> At the end of it, index() is currently nearly 100 lines but can be just
> this:
>
> @app.route('/', methods = ["GET"])
> def index():
>     query = {
>         'key': trellokey,
>         'token': trellotoken
>     }
>     cards_response = requests.request(
>         "GET",
>         boardurl,
>         params=query
>     )
>     cards = json.loads(cards_response.text)
>     todos = []
>     for card in cards:
>         todos.append(card["name"])
>
>     return render_template('index.html', todisplay=todos)
>
> All I've done is rename some variables and delete comments and code that
> shouldn't be there. It can be tidied up further, but this is good enough.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475738918>
> :
>
> >      return redirect("/")
>
> +@app.route('/complete_item', methods = ["PUT","GET","POST"])
> +
> +def complete_item():
> +
> +    query = {
> +        'key': trellokey,
> +        'token': trellotoken,
> +        'idList': donelistid,
> +        'daveurl': "https://api.trello.com/1/cards" + request.form['title2']
>
> This line should be deleted, right?
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475739210>
> :
>
> >      return redirect("/")
>
> +@app.route('/complete_item', methods = ["PUT","GET","POST"])
> +
> +def complete_item():
> +
> +    query = {
> +        'key': trellokey,
> +        'token': trellotoken,
> +        'idList': donelistid,
> +        'daveurl': "https://api.trello.com/1/cards" + request.form['title2']
> +    }
> +    print(query)
> +    response = requests.request(
> +        "PUT",
> +        "https://api.trello.com/1/cards/" + request.form['title2'],                 # The API demands the number is put into the URL, not as part of the query
>
> It should be called id, not "title2"
> ------------------------------
>
> In templates/index.html
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475747354>
> :
>
> >        </ul>
> +
> +
> +      <h2>Items from TRELLO To-DO list</h2>
> +      <ul class="list-group mb-4">
> +        <!--Literally no for loop would work.  After 4 hours of trying, It seems impossible to display a passed Python list as HTML list - so I had no choice but to do the below-->
> +
> +        {% for todo in todisplay %}
> +          <li>{{todo}}</li>
> +          <button class="button button1">Move to complete</button>
>
> This should be a <form> with a hidden input set to the todo's id, and a
> submit button. The submit button is all the user sees.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475750092>
> :
>
> > +    # Titleback=request.form.get('title')
> +    # Lets try putting the Trello INSERT CARD thing in here.  First QUERY is what we're gonna send to the API
>
> These two comment lines should be deleted.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475763535>
> :
>
> > + #   print(response2.text)                           # Just to check its actually being gotten!
> +
> +    #print(response2.text)
> +    #print(type(response2.text))                     # OK so response2.text is one massive long string.  Which is fairly useless
> +                                                    # It was a nightmare getting the bit of the JSON I wanted
> +    the_list = json.loads(response2.text)           # transfer the one long string into a LIST (Not a dictionary .. pretty useless now I'm gonna have to split it by hand). I REALLY with I knew how to get it into a dictionery, would have saved hours.
> +#    firstcard=str(card_data[1])
> +#    nowsplit=firstcard.split()
> +    #print (nowsplit['Id'])
> + #   thetype = type(card_data)
> +    #  print(the_list[1])                                 # Here for example is the first card details.  But still, this massive list is actually simply 1 element of a string, which is no use
> +                                                    # So now I need to think of some way of getting this single string into seperate elements manually
> +    for todo in the_list:
> +        superlist.append(todo["name"])                              # for each of the full cards in the list
> +
> +# I tried all the below but literally nothing would work in order to get those card id's into different parts of the list, so next tried this:
>
> Sorry, not sure what you wanted to achieve. I'm happy to discuss it if you
> had something particular in mind that you wanted to do but couldn't work
> out how to do it in Python.
> ------------------------------
>
> In app.py
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#discussion_r475749815>
> :
>
> > +
> +    # During module_1 a dict was returned that was in a format that somehow HTML seems to deal with.  No matter what I try I can't get 'superlist' into a format that simulates the format of the cookies list that works on the HTML.  It's too hard.  I've spent 4 hours on it.  It seems impossible with the info we have.
> +    # I'm having to very deep dive the 'enumerate' and indices operations but I can't spend lots more hours in this.
> +    #superlist2 = {key: [index for index, x in enumerate(superlist) if x == key] for key in superlist}.  All the below fails.
> +    #my_dict={
> +    #for key in superlist:
> +    #    indices = []
> +    #    for index,value in enumerate(superlist):
> +    #        if value == key:
> +    #            indices.append(index)
> +    #    my_dict[key] = indices
> +    #print(my_dict)
> +
> +    # print(superlist2)
> +
> +    return render_template('index.html',passedItems=Items,todisplay=superlist)
>
> To support having button on the page for updating the todo, you need to
> pass both the id and the title to the template. So you should build a list
> of objects rather than just a list of strings.
>
> The exercise brief tells you to define an Item class. And then you would
> change the append line to something like: todos.append(Item(card["id"],
> card["name"]) and in index.html you would display it with something like {{
> todo.title }}, depending on how you defined the Item class.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/britboy4321/DevOps-Course-Starter/pull/3#pullrequestreview-473614698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQI6L6TVHTPWAK2IFHCJGILSCKPGFANCNFSM4QDXYBVA>
> .
>

hugh-emerson commented on .vscode/settings.json:
```diff
@@ -0,0 +1,9 @@
+{
+    "python.pythonPath": "c:\\Users\\Dave\\DevOps-Course-Starter\\env\\Scripts\\python.exe",
```
This should be a relative path... unless everyone the team is called Dave

It can just be "env\\Scripts\\python.exe"

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+    thislist=[]                  
+    superlist=[] 
+    cardsurl = "https://api.trello.com/1/cards"      
+    boardurl = "https://api.trello.com/1/boards/5f3528983d4fb244aae9f934/cards"             # The board ID is not a secret!
+    listid = "5f352898dc8a8c31a0a1e439"                         
```
This works if I replace these ids and donelistid inside the complete_item() function but it would be nicer to include them in .env as well, just like your "key" and "token" environment variables.


And then here you can do `listid=os.environ["TODO_LIST_ID"]` and `boardurl=f"...trellostuff/{os.environ["BOARD_ID"]/cards}"`

Your solution works for now but you will need this change for the module 3 exercise

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
-    Items=session.get_items()   
-   # session.add_item('A super entry')   # This works to add entries
-    return render_template('index.html',passedItems=Items)
+    Items=session.get_items()
```
Can you delete the session_items.py file and everything related to it? So this line, the import, anything to do with "passedItems" here and in index.html

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+    query = {
+        'key': trellokey,
+        'token': trellotoken,
+        'idList': listid,
```
`listid`  and `cardsurl` are undefined since you moved them.

(Also listid should probably be renamed to indicate it's specifically for "to do" items, not completed items.)

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+
+trellokey=os.environ["key"]            # get the secret key
+trellotoken=os.environ["token"]         # get the secret token
+@app.route('/', methods = ["GET","PUT"])
```
methods should just be `= ["GET"]`

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+    )
+    return redirect("/")
+
+@app.route('/complete_item', methods = ["PUT","GET","POST"])
```
methods should just be `["POST"]`

hugh-emerson commented on templates/index.html:
```diff
@@ -9,20 +9,72 @@ <h1 class="display-4">To-Do App</h1>
+      </ul>  
+      
+      <script> 
+        function function1() { 
+        document.body.style.backgroundColor = 'blue'; 
+        } 
+        function function2() { 
+        document.body.style.backgroundColor = 'red';  
+        } 
+        function function3() { 
+        document.body.style.backgroundColor = 'green';  
+        }  
```
Unused

hugh-emerson commented on templates/index.html:
```diff
@@ -9,20 +9,72 @@ <h1 class="display-4">To-Do App</h1>
     <input type="text" id="title" name="title"><br>
   
   </form>
+  <form action="/complete_item" method="POST">
+    <label for="title">CARD ID of item to be moved to DONE:</label><br>
+    <input type="text" id="title2" name="title2"><br>
+  </form>
```
You can delete this form now

hugh-emerson commented on templates/index.html:
```diff
@@ -9,20 +9,72 @@ <h1 class="display-4">To-Do App</h1>
+      <ul class="list-group mb-4">
+       
+        {% for todo in todisplay %}     
+          <li>{{todo.name}}</li>
```
It would be very helpful to include a status... you currently can't see what's "to do" and what's "completed".

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+    the_list = json.loads(board_response.text)     # A list of cards
+    for todo in the_list:
+       
+        superlist.append({'name': todo['name'], 'id': todo['id']})
```
Part 5 of the exercise brief asks you to create a `Class`

You define a class like so:
```
class Item:
    def __init__(self, name, id, status):
        self.name = name
        self.id = id
        self.status = status
```

And then instead of appending a dictionary `{'name': todo['name'], 'id': todo['id']}`

You should `.append(Item(todo['name'], todo['id'], status)`

Currently you're not including the item status... to do so you can do something like:
```
if todo['idList'] == donelistid:
    status = 'Done'
else:
    status = 'To Do'
```
If you uncomment donelistid. Or the reverse with listid... but you should rename it to indicate it's the todo list.

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
+    Items=session.get_items()
+    thislist=[]                  
+    superlist=[] 
+    cardsurl = "https://api.trello.com/1/cards"      
```
`cardsurl` and `listid` only exist inside index() at the moment but it's entry() that wants them

hugh-emerson commented on app.py:
```diff
@@ -1,22 +1,81 @@
-   # session.add_item('A super entry')   # This works to add entries
-    return render_template('index.html',passedItems=Items)
+    Items=session.get_items()
+    thislist=[]                  
```
Unused

hugh-emerson reviewed (approved):
This is almost there, and I'm marking it as approved but technically there are two small ways that you don't meet the minimum requirements.

1 - adding cards no longer works since your latest changes because you moved where some variables were defined
2 - the exercise brief asks you to define a class for your todo items, which you have not attempted.

See my comments for more explanation. There are other suggestions as well but those two are the most important.
